frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/index.tsx

```diff
@@ -0,0 +1,113 @@
+import * as React from 'react';
+import * as _ from 'lodash';
+import { NodeModel } from '@console/internal/models';
+import { createModalLauncher, ModalComponentProps } from '@console/internal/components/factory';
+import {
+  withHandlePromise,
+  HandlePromiseProps,
+  Firehose,
+} from '@console/internal/components/utils';
+import { k8sPatch } from '@console/internal/module/k8s';
+import { getVMLikeModel, getNodeSelector } from '../../../../selectors/vm';
+import { getNodeSelectorPatches } from '../../../../k8s/patches/vm/vm-scheduling-patches';
+import { VMLikeEntityKind } from '../../../../types/vmLike';
+import { NodeSelector } from '../../../../types';
+import { IDLabel } from '../shared';
+import { NSModal } from './node-selector-modal';
+import { getName, getNamespace } from '@console/shared';
+
+const NodeSelectorModal: React.FC<NodeSelectorModalProps> = (props) => {
+  const { vmLikeEntity, nodeSelector = {}, handlePromise, cancel, close, ...restProps } = props;
+
+  const [initialNodeSelector, setInitialNodeSelector] = React.useState<NodeSelector>(nodeSelector);
+  const [selectorLabels, setSelectorLabels] = React.useState<IDLabel[]>(
+    Object.entries(nodeSelector).map(([key, value], id) => ({ id, key, value })),
+  );
+
+  const onReload = (vm: VMLikeEntityKind) => {
+    const updatedSelector = getNodeSelector(vm);
+    setInitialNodeSelector(updatedSelector);
+    setSelectorLabels(
+      Object.entries(updatedSelector || {}).map(([key, value], id) => ({ id, key, value })),
+    );
+  };
+
+  const onLabelAdd = () => {
+    const id = selectorLabels[selectorLabels.length - 1]?.id + 1 || 0;
+    setSelectorLabels([...selectorLabels, { id, key: '', value: '' }]);
+  };
+
+  const onLabelChange = (newLabel: IDLabel) =>
+    setSelectorLabels(
+      selectorLabels.map((label) => {
+        if (label.id === newLabel.id) {
+          return newLabel;
+        }
+        return label;
+      }),
+    );
+
+  const onLabelDelete = (idToDelete) => {
+    setSelectorLabels(selectorLabels.filter(({ id }) => id !== idToDelete));
+  };
+
+  const onSubmit = async () => {
+    const k8sSelector = Object.assign(
+      {},
+      ...selectorLabels.filter(({ key }) => !!key).map(({ key, value }) => ({ [key]: value })),
+    );
+
+    if (!_.isEqual(nodeSelector, k8sSelector)) {
+      // eslint-disable-next-line promise/catch-or-return
+      handlePromise(
+        k8sPatch(
+          getVMLikeModel(vmLikeEntity),
+          vmLikeEntity,
```

Comment:
vmLikeFinal should be used in the submit request as well - or we might reach inconsistencies

Reply:
It would be probably much easier to keep the initialNodeSelector state and the callback inside NSModal and use this one only as a resource wrapper - so we will have the recent VM in the scope